csrc/fusion_profiler.cpp

```diff
@@ -0,0 +1,737 @@
+           << ", " << std::get<2>(kp.grid) << "]";
+      std::stringstream block;
+      block << "[" << std::get<0>(kp.block) << ", " << std::get<1>(kp.block)
+            << ", " << std::get<2>(kp.block) << "]";
+      std::stringstream cluster;
+      cluster << "[" << std::get<0>(kp.cluster) << ", "
+              << std::get<1>(kp.cluster) << ", " << std::get<2>(kp.cluster)
+              << "]";
+      std::stringstream smem;
+      smem << "[" << kp.dynamic_shared_mem << ", " << kp.static_shared_mem
+           << "]";
+      os << std::setfill(' ') << std::right << std::fixed << " " << std::setw(4)
+         << idx << " " << std::setw(10) << kp.name << " " << std::setw(11)
+         << std::setprecision(3) << kp.time_ms;
+
+      if (fp.verbose) {
+        os << " " << std::setw(11) << std::setprecision(3)
+           << kp.compile_time_ms;
+      }
+
+      os << " " << std::setw(13) << std::setprecision(2)
+         << kp.effective_bandwidth_gbs << " " << std::setw(9)
+         << std::setprecision(2) << kp.percentage_peak_bandwidth << " "
+         << std::setw(9) << std::setprecision(3)
+         << ((double)kp.input_bytes / 1000000.0) << " " << std::setw(9)
+         << std::setprecision(3) << ((double)kp.output_bytes / 1000000.0) << " "
+         << std::setw(16) << smem.str() << " " << std::setw(6) << kp.registers
+         << " " << std::setw(16) << grid.str() << " " << std::setw(16)
+         << block.str();
+      if (fp.verbose) {
+        os << " " << std::setw(16) << cluster.str() << " " << std::setw(5)
+           << kp.device << " " << std::setw(5) << kp.stream << " "
+           << std::setw(20) << kp.device_name << " " << std::setw(14)
+           << std::setprecision(2) << kp.peak_bandwidth_gbs;
+      }
+      os << std::endl;
+      ++idx;
+    }
+  }
+  return os;
+}
+
+std::mutex FusionProfiler::singleton_lock_;
+FusionProfiler* FusionProfiler::singleton_ = nullptr;
+
+FusionProfiler::FusionProfiler()
+    : cupti_disabled_(false),
+      cupti_buffer_((size_t)4 * 1024),
+      state_(ProfilerState::Ready),
+      fusion_id_(-1),
+      parallel_compile_(false),
+      profile_(),
+      fusion_timer_(at::cuda::getCurrentCUDAStream()),
+      host_timer_(),
+      parallel_compile_timer_(),
+      segments_(),
+      device_descriptors_(),
+      kernel_profiles_(),
+      corrid_2_segid_() {
+  if (!cupti_disabled_) {
+    NVFUSER_CUPTI_SAFE_CALL(cuptiActivityRegisterCallbacks(
+        cupti_buffer_requested, cupti_buffer_completed));
+  }
+}
+
+void FusionProfiler::reset() {
+  state_ = ProfilerState::Ready;
+  ++fusion_id_;
+
+  profile_.reset();
+  fusion_timer_.reset();
+  host_timer_.reset();
+  parallel_compile_ = false;
+  parallel_compile_timer_.reset();
+  segments_.clear();
+  kernel_profiles_.clear();
+  corrid_2_segid_.clear();
+}
+
+FusionProfiler* FusionProfiler::get() {
+  std::lock_guard<std::mutex> guard(singleton_lock_);
+  if (singleton_ == nullptr) {
+    singleton_ = new FusionProfiler();
```

Comment:
nitpick: maybe std::unique_ptr instead.

Reply:
A static std::unique_ptr? There's a good reason not to do that: https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables

Reply:
In the same spirit, I think we should move singleton_lock_ to this function as a local variable so the time of construction is predictable and thread-safe.

Reply:
I made the suggested change.